merlin-driver/src/main/java/gov/nasa/jpl/aerie/merlin/driver/CheckpointSimulationDriver.java

```diff
@@ -0,0 +1,412 @@
+
+    // Specify a topic on which tasks can log the activity they're associated with.
+
+    try {
+      final var filteredSchedule = new HashMap<ActivityDirectiveId, ActivityDirective>();
+      for (final var entry : schedule.entrySet()) {
+        if (entry.getValue().startOffset().longerThan(cachedEngine.endsAt())) {
+          filteredSchedule.put(entry.getKey(), entry.getValue());
+        }
+      }
+
+      // Get all activities as close as possible to absolute time
+      // Schedule all activities.
+      // Using HashMap explicitly because it allows `null` as a key.
+      // `null` key means that an activity is not waiting on another activity to finish to know its start time
+      HashMap<ActivityDirectiveId, List<Pair<ActivityDirectiveId, Duration>>> resolved = new StartOffsetReducer(planDuration, filteredSchedule).compute();
+      if(!resolved.isEmpty()) {
+        resolved.put(
+            null,
+            StartOffsetReducer.adjustStartOffset(
+                resolved.get(null),
+                Duration.of(
+                    planStartTime.until(simulationStartTime, ChronoUnit.MICROS),
+                    Duration.MICROSECONDS)));
+      }
+      // Filter out activities that are before simulationStartTime
+      resolved = StartOffsetReducer.filterOutNegativeStartOffset(resolved);
+
+      final var toSchedule = new LinkedHashSet<ActivityDirectiveId>();
+      toSchedule.add(null);
+      var toCheckForDependencyScheduling = scheduleActivities(
+          toSchedule,
+          filteredSchedule,
+          resolved,
+          missionModel,
+          engine,
+          elapsedTime,
+          activityToSpan,
+          activityTopic);
+
+      // Drive the engine until we're out of time.
+      // TERMINATION: Actually, we might never break if real time never progresses forward.
+      while (elapsedTime.noLongerThan(simulationDuration) && !simulationCanceled.get()) {
+        final var nextTime = engine.peekNextTime().orElse(Duration.MAX_VALUE);
+        if (shouldTakeCheckpoint.apply(new SimulationState(elapsedTime, nextTime, engine, schedule, activityToSpan))) {
+          if(duplicationIsOk) cells.freeze();
+          LOGGER.info("Saving a simulation engine in memory at time " + elapsedTime + " (next time: " + nextTime + ")");
+          final var newCachedEngine = new CachedSimulationEngine(
+              elapsedTime,
+              schedule,
+              engine,
+              cells,
+              makeCombinedTimeline(timelines, timeline).points(),
+              activityTopic,
+              missionModel);
+          if(duplicationIsOk) newCachedEngine.freeze();
+          cachedEngineStore.save(
+              newCachedEngine,
+              configuration);
+          timelines.add(timeline);
+          engine = duplicationIsOk ? engine.duplicate() : engine;
+          timeline = new TemporalEventSource();
+          cells = new LiveCells(timeline, cells);
+        }
```

Comment:
It won't be modified because the only checkpoint that will be taken in that case is when it's about to stop simulation. Look at line 230 of `CheckpointSimulationFacade`.

Reply:
note: two branches depending on the behavior